File: jvm/selfie-lib/src/commonMain/kotlin/com/diffplug/selfie/guts/WithinTestGC.kt
/*
 * Copyright (C) 2024 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.selfie.guts

import com.diffplug.selfie.ArrayMap
import com.diffplug.selfie.ArraySet
import com.diffplug.selfie.Snapshot

/** Handles garbage collection of snapshots within a single test. */
class WithinTestGC {
  private val suffixesToKeep: AtomicRef<ArraySet<String>?> = atomic(ArraySet.empty())
  fun keepSuffix(suffix: String) {
    suffixesToKeep.updateAndGet { it?.plusOrThis(suffix) }
  }
  fun keepAll(): WithinTestGC {
    suffixesToKeep.updateAndGet { null }
    return this
  }
  override fun toString() = suffixesToKeep.get()?.toString() ?: "(null)"
  fun succeededAndUsedNoSnapshots() = suffixesToKeep.get() === ArraySet.empty<String>()
  private fun keeps(s: String): Boolean = suffixesToKeep.get()?.contains(s) ?: true

  companion object {
    fun findStaleSnapshotsWithin(
        snapshots: ArrayMap<String, Snapshot>,
        testsThatRan: ArrayMap<String, WithinTestGC>,
        testsThatDidntRun: Sequence<String>
    ): List<Int> {
      val staleIndices = mutableListOf<Int>()

      // - Every snapshot is named `testMethod` or `testMethod/subpath`
      // - It is possible to have `testMethod/subpath` without `testMethod`
      // - If a snapshot does not have a corresponding testMethod, it is stale
      // - If a method ran successfully, then we should keep exclusively the snapshots in
      // MethodSnapshotUsage#suffixesToKeep
      // - Unless that method has `keepAll`, in which case the user asked to exclude that method
      // from pruning

      // combine what we know about methods that did run with what we know about the tests that
      // didn't
      var totalGc = testsThatRan
      for (method in testsThatDidntRun) {
        totalGc = totalGc.plus(method, WithinTestGC().keepAll())
      }
      val gcRoots = totalGc.entries
      val keys = snapshots.keys
      // we'll start with the lowest gc, and the lowest key
      var gcIdx = 0
      var keyIdx = 0
      while (keyIdx < keys.size && gcIdx < gcRoots.size) {
        val key = keys[keyIdx]
        val gc = gcRoots[gcIdx]
        if (key.startsWith(gc.key)) {
          if (key.length == gc.key.length) {
            // startWith + same length = exact match, no suffix
            if (!gc.value.keeps("")) {
              staleIndices.add(keyIdx)
            }
            ++keyIdx
            continue
          } else if (key.elementAt(gc.key.length) == '/') {
            // startWith + not same length = can safely query the `/`
            val suffix = key.substring(gc.key.length)
            if (!gc.value.keeps(suffix)) {
              staleIndices.add(keyIdx)
            }
            ++keyIdx
            continue
          } else {
            // key is longer than gc.key, but doesn't start with gc.key, so we must increment gc
            ++gcIdx
            continue
          }
        } else {
          // we don't start with the key, so we must increment
          if (gc.key < key) {
            ++gcIdx
          } else {
            // we never found a gc that started with this key, so it's stale
            staleIndices.add(keyIdx)
            ++keyIdx
          }
        }
      }
      while (keyIdx < keys.size) {
        staleIndices.add(keyIdx)
        ++keyIdx
      }
      return staleIndices
    }
  }
}

File: python/selfie-lib/selfie_lib/WithinTestGC.py
from collections.abc import Iterable
from threading import Lock

from .ArrayMap import ArrayMap, ArraySet
from .Snapshot import Snapshot


class WithinTestGC:
    def __init__(self):
        self.suffixes_to_keep = ArraySet.empty()
        self.lock = Lock()

    def keep_suffix(self, suffix: str):
        with self.lock:
            if self.suffixes_to_keep:
                self.suffixes_to_keep = self.suffixes_to_keep.plusOrThis(suffix)

    def keep_all(self) -> "WithinTestGC":
        with self.lock:
            self.suffixes_to_keep = None
        return self

    def __str__(self) -> str:
        with self.lock:
            return (
                str(self.suffixes_to_keep)
                if self.suffixes_to_keep is not None
                else "(null)"
            )

    def succeeded_and_used_no_snapshots(self) -> bool:
        with self.lock:
            return self.suffixes_to_keep == ArraySet.empty()

    def keeps(self, s: str) -> bool:
        with self.lock:
            return True if self.suffixes_to_keep is None else s in self.suffixes_to_keep

    @staticmethod
    def find_stale_snapshots_within(
        snapshots: ArrayMap[str, Snapshot],
        tests_that_ran: ArrayMap[str, "WithinTestGC"],
        tests_that_didnt_run: Iterable[str],
    ) -> list[int]:
        stale_indices = []

        # combine what we know about methods that did run with what we know about the tests that didn't
        total_gc = tests_that_ran
        for method in tests_that_didnt_run:
            total_gc = total_gc.plus(method, WithinTestGC().keep_all())

        gc_roots = total_gc.items()
        keys = snapshots.keys()
        # we'll start with the lowest gc, and the lowest key
        gc_idx = 0
        key_idx = 0
        while key_idx < len(keys) and gc_idx < len(gc_roots):
            key: str = keys[key_idx]  # type: ignore
            gc: tuple[str, WithinTestGC] = gc_roots[gc_idx]  # type: ignore
            if key.startswith(gc[0]):
                if len(key) == len(gc[0]):
                    # startWith + same length = exact match, no suffix
                    if not gc[1].keeps(""):
                        stale_indices.append(key_idx)
                    key_idx += 1
                    continue
                elif key[len(gc[0])] == "/":
                    # startWith + not same length = can safely query the `/`
                    suffix = key[len(gc[0]) :]
                    if not gc[1].keeps(suffix):
                        stale_indices.append(key_idx)
                    key_idx += 1
                    continue
                else:
                    # key is longer than gc.key, but doesn't start with gc.key, so we must increment gc
                    gc_idx += 1
                    continue
            else:
                # we don't start with the key, so we must increment
                if gc[0] < key:
                    gc_idx += 1
                else:
                    # we never found a gc that started with this key, so it's stale
                    stale_indices.append(key_idx)
                    key_idx += 1

        while key_idx < len(keys):
            stale_indices.append(key_idx)
            key_idx += 1

        return stale_indices
